File: examples/concurrency/threads.cpp
#include <iostream>
#include <thread>
#include <vector>
#include <atomic>
#include <cmath>

// Global stop signal
std::atomic<bool> stop_requested(false);

bool is_prime(int num) {
    if (num <= 1) return false;
    for (int i = 2; i <= std::sqrt(num); ++i) {
        if (num % i == 0) return false;
    }
    return true;
}

void find_primes(std::atomic<int>& counter) {
    int num = 2;
    while (!stop_requested.load()) {
        if (is_prime(num)) {
            std::cout << "Prime found: " << num << std::endl;
            counter.fetch_add(1, std::memory_order_relaxed);
        }
        num += 1;  // Move to the next number
    }
}

int main() {
    std::cout << "Press Enter to start...\n";
    std::cin.get();
    int num_threads = 4;
    std::vector<std::thread> threads;
    std::atomic<int> prime_count = 0;

    // Start prime-finding threads
    for (int i = 0; i < num_threads; ++i) {
        threads.push_back(std::thread(find_primes, std::ref(prime_count)));
    }

    // Wait for a stop command
    std::cout << "Press Enter to stop...\n";
    std::cin.get();
    stop_requested.store(true);

    // Join all threads
    for (auto& thread : threads) {
        thread.join();
    }
    std::cout << "Press Enter to exit...\n";
    std::cin.get();
    return 0;
}

// Compile with g++ -pthread threads.cpp

File: examples/concurrency/threads.py
import threading

def find_primes(stop_event):
    num = 2
    while not stop_event.is_set():
        if is_prime(num):
            print(f"Prime found: {num}")
        num += 1

def is_prime(num):
    """Returns True if num is a prime number, otherwise False."""
    if num < 2:
        return False
    for i in range(2, int(num ** 0.5) + 1):
        if num % i == 0:
            return False
    return True

def main():
    input("Press Enter to start...\n")
    # Create an event to signal the thread to stop
    stop_event = threading.Event()

    # Start the prime-finding threads
    threads = []
    for _ in range(4):
        thread = threading.Thread(target=find_primes, args=(stop_event,))
        thread.start()
        threads.append(thread)

    try:
        # Keep the main program running for a few seconds
        input("Press Enter to stop...\n")
    finally:
        # Set the event to stop the threads
        stop_event.set()
        for thread in threads:
            thread.join()   # Wait for the prime threads to finish

    input("Press Enter to exit...\n")

if __name__ == "__main__":
    main()